website/blog/2023-02-10.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,605 @@
+* Empty array is less than null.
+
+### Sorting Array
+Let's insert following records to see how arrays are sorted.
+
+```js
+db.mixedsort.insertMany([
+{ _id: 'array-empty', v: [] },
+{ _id: 'null', v: null },
+{ _id: 'array-mixed', v: [ 'a', 'z', 0 ] },
+{ _id: 'array-numbers', v: [ 10, 5 ] },
+{ _id: 'array-strings', v: [ 'a' ] },
+{ _id: 'string', v: 'b' }
+])
+```
+
+Let's use ascending sort on `v`, ascending sort is indicated by `1`.
+
+```js
+db.mixedsort.find().sort({v: 1})
+[
+  { _id: 'array-empty', v: [] },              // empty array is the smallest
+  { _id: 'null', v: null },                   // null is used for comparison
+  { _id: 'array-mixed', v: [ 'a', 'z', 0 ] }, // 0 the smallest value is used for comparison
+  { _id: 'array-numbers', v: [ 10, 5 ] },     // 5 the smallest value is used for comparison
+  { _id: 'array-strings', v: [ 'a' ] },       // 'a' the smallest value is used for comparison
+  { _id: 'string', v: 'b' }                   // 'b' is used for comparison
+]
+```
+Ascending sort uses the smallest element from array for comparison.
+For example, `array-mixed` record gets its smallest element `0` and uses
+that to compare against other records.
+Even though an array is greater than string in [BSON comparison order](#bson-comparison-order),
+`array-mixed` record is smaller than
+`string` record because the smallest element `'a'` from `array-mixed` and `'b'` from `string`
+are compared.
+
+
+Let's use descending sort on `v`, descending sort is indicated by `-1`.
+
+```js
+db.mixedsort.find().sort({v: -1})
+[
+  { _id: 'array-mixed', v: [ 'a', 'z', 0 ] }, // 'z' the greatest value is used for comparison
+  { _id: 'string', v: 'b' },                  // 'b' is used for comparison
+  { _id: 'array-strings', v: [ 'a' ] },       // 'a' the greatest value is used for comparison
+  { _id: 'array-numbers', v: [ 10, 5 ] },     // 10 the greatest value is used for comparison
+  { _id: 'null', v: null },                   // null is used for comparison
+  { _id: 'array-empty', v: [] }               // empty array is the smallest
+]
+```
+
+Descending sort uses the largest element of the array.
+For example, `array-mixed` record gets its largest element `'z'` and uses
+that to compare against other records.
+
+Descending sort is not the opposite order of ascending sort,
+because the comparison element used from the array is different.
+
+### Sorting Nested Array
+
+Let's use following example to sort nested arrays.
+```js
+ db.nestedsort.insertMany([
+  { _id: 'nested-number', v: [ [ 6, 1 ] ] },
+  { _id: 'nested-null', v: [ ['g'], [ null ] ] },
+  { _id: 'nested-mixed', v: [ [ true, 0, 'g' ] ] },
+  { _id: 'nested-empty', v: [ [ 6, 2 ], [ ] ] }, 
+  { _id: 'array-mixed', v: [ true, 0, 'g' ] }
+])
+```
+
+Like [array](#sorting-array) nested array gets the largest or smallest element
+from the array, but for the case of nested array it's the smallest inner array.
```

Comment:
I don't really understand this sentence. Probably needs a rewrite.

Reply:
Rewritten this part :hugs: